Sources/XCTest/Public/XCTestCase.swift

```diff
@@ -195,66 +286,40 @@ open class XCTestCase: XCTest {
     /// class.
     open class func tearDown() {}
 
-    private var teardownBlocks: [() -> Void] = []
-    private var teardownBlocksDequeued: Bool = false
-    private let teardownBlocksQueue: DispatchQueue = DispatchQueue(label: "org.swift.XCTest.XCTestCase.teardownBlocks")
+    private actor TeardownBlocksState {
+        
+        private var wasFinalized = false
+        private var blocks: [() async throws -> Void] = []
```

Comment:
I just realized that having this `Array` of `addTeardownBlock` closures include the `async` keyword will break existing behavior by executing **all** the closures via `Task {}`, because that will mean that existing, non-async teardown blocks will execute off the main thread and that could break assumptions. (This is a similar problem as the one we faced for XCTestCase closures themselves.)

I think we could employ a similar solution here of having this Array store instance of that `enum Closure` introduced elsewhere in the PR, and conditionally executing each teardown block as either `.sync` or `.async`

Reply:
Ah right! Great memory. We had discussed this earlier in the PR's history on Slack. I'll add that change in.

Reply:
So, `TeardownBlocksState` needs to be a `class` and not an `actor` so we can make sure that teardown blocks are added to our stack from the correct thread. I've reverted the kind of type it is and used the existing `XCTWaiter.subsystemQueue.sync` method to synchronize adding to the teardown block stack. After doing this, I was able to make a new `enum` called `XCTestCase.TeardownBlocksState.BlockType` which affords us the same flexibility as the newly introduced `XCTestCase.ClosureType`. 

The reason I created a new `enum` is because teardown blocks have a different function signatures than the `XCTestCase.ClosureType` enum, and it seems like a bad design to need to figure out when to deal with some cases and not the other based on the usage site. The best thing is a new type that's used identically, but returns `Void` and is argumentless,.